interpreter/attribute_patterns_test.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/google/cel-go/common/packages"
+	"github.com/google/cel-go/common/types"
+)
+
+type attr struct {
+	maybe bool
+	name  string
+	quals []interface{}
+}
+
+type patternTest struct {
+	pattern *AttributePattern
+	pkg     string
+	matches []attr
+	misses  []attr
+}
+
+var patternTests = map[string]patternTest{
+	"var": {
+		pattern: NewAttributePattern("var"),
+		matches: []attr{
+			{name: "var"},
+			{name: "var", quals: []interface{}{"field"}},
+		},
+		misses: []attr{
+			{name: "ns.var"},
+		},
+	},
+	"var_namespace": {
+		pattern: NewAttributePattern("ns.app.var"),
+		pkg:     "ns.app",
+		matches: []attr{
+			{name: "ns.app.var"},
+			{name: "ns.app.var", quals: []interface{}{int64(0)}},
+			{name: "ns", quals: []interface{}{"app", "var", "foo"}, maybe: true},
```

Comment:
Ah, this is a `MaybeAttribute` which represents an unchecked expression whose checked form might be any one of the following:

`id: ns.app.var.foo`
`select(id: ns.app.var, field: 'foo')`
`select(select(id: ns.app, field: 'var'), field: 'foo')`
`select(select(select(id: ns, field: 'app'), field: 'var'), field: 'foo')`

The test ensures that even parsed only attributes will match the pattern correctly.

Reply:
So it seems that the container setting isn't really relevant to the test - maybe take it out?  Or make it relevant by dropping the "ns.app" prefix from one of the cases?

But again with Issue 53:  it seems like we're declaring that "ns.app.var" is a real variable.  We might decide that this forbids us from having "ns" or "ns.app" as variables.

At least put in the following comment:
```
// Note that https://github.com/google/cel-spec/issues/53 complains about
// variable names that are a dotted prefix of other qualified variable names,
// so this tests some scenarios which might be disallowed in the future.
```

Reply:
However, this test is not exercising subsumption, but rather demonstrating that maybe resolution considers the container when attempting to match a parsed-only expression against the attribute pattern. In the examples I typed out earlier, maybe option number 2 is the one that matches.

Reply:
I've added a comment about how 'maybe' fits into the test. Perhaps I should describe the test-case a bit differently.

Reply:
Per our discussion, I see that the matching "ns" plus qualifiers doesn't imply a variable named "ns", since the "unchecked" bit is set, so no warning vs Issue 53 is needed.